File: python/inversion_GD.py
#############################################################################
#    INVERSION CODE EXAMPLE USING FORWARD_SOLVER MODEL ATMOSPHERE           #
#                AUTHOR: ANDRES VICENTE AREVALO                             #
#############################################################################
import numpy as np
import matplotlib.pyplot as plt
import forward_solver_py as fs
import parameters as pm

##########################     SUBROUTINES     ##############################
def add_noise(array, sigma):
    """
    Add nromal noise to an N dimensional array with a ginven sigma
    """
    noise = np.random.normal(0,sigma,array.size)
    noise = noise.reshape(array.shape)
    array = array + noise
    return array


def chi2(params, I_obs_sol, Q_obs_sol, std, w_I, w_Q, mu):
    '''
    Compute the cost function of the inversion given the parameters, the noise
    and weigths of the diferent components (I,Q...)
    '''
    a, r, eps, dep_col, Hd = params[0], params[1], params[2], params[3], params[4]

    I, Q = fs.solve_profiles(a, r, eps, dep_col, Hd)

    I_obs = I[-1,:,mu].copy()
    Q_obs = Q[-1,:,mu].copy()

    chi2 = np.sum(w_I*(I_obs-I_obs_sol)**2/std**2 + w_Q*(Q_obs-Q_obs_sol)**2/std**2)/(2*I_obs.size)

    print(f'Total Chi^2 of this profiles is: {chi2}')
    return chi2, I_obs, Q_obs #, Jm00_new, Jm20_new


def surroundings(x_0, h):

    surroundings = np.zeros((x_0.shape[0], *x_0.shape))
    for i in range(x_0.shape[0]):
        delta = np.zeros_like(x_0)
        delta[i] = delta[i] + h
        surroundings[i] = x_0 + delta

    return surroundings


def compute_gradient(I_sol, Q_sol, I_0, Q_0, x_0, xs, std, w_I, w_Q, mu):

    chi2_pivot, _ , _  = chi2(x_0, I_sol, Q_sol, std, w_I, w_Q, mu)
    chi2s = np.ones((x_0.shape[0],2))
    chi2s[:,0] = np.ones(x_0.shape[0])*chi2_pivot
    
    for i in range(len(x_0)):
        chi2s[i,1], _ , _ = chi2(xs[i], I_sol, Q_sol, std, w_I, w_Q, mu)

    _ , beta = np.gradient(chi2s,1)
    beta = beta[:,0]

    return beta

####################    COMPUTE THE "OBSERVED" PROFILE    #####################
a_sol = 1e-3      #1e-3                 # dumping Voigt profile a=gam/(2^1/2*sig)
r_sol = 1e-2      #1e-2                 # XCI/XLI
eps_sol = 1e-1    #1e-1                 # Phot. dest. probability (LTE=1,NLTE=1e-4)
dep_col_sol = 8   #8                    # Depolirarization colisions (delta)
Hd_sol = .8       #.8                  # Hanle depolarization factor [1/5, 1]
mu = 9 #int(fs.pm.qnd/2)

##############      INITIALICE THE PARAMETERS       #######################
seed = 666
np.random.seed(seed)
a_initial =  10**(-np.random.uniform(0,10))
r_initial =  10**(-np.random.uniform(0,12))
eps_initial = 10**(-np.random.uniform(0,4))
dep_col_initial =  np.random.uniform(0,1)
Hd_initial =  np.random.uniform(1/5, 1)

w_I     , w_Q   = 1e-1  , 1e3

h = 1e-8
max_itter = 250
std = 1e-5
# step_size = 1e-5
step_size = np.array([1e-4,1e-4,1e-4,1e0,1e0])


###################     COMPUTE THE SOLUTION PROFILES AND RADIATION FIELD       ###################
print("Solution parameters: ")
print(f" a = {a_sol}\n r = {r_sol}\n eps = {eps_sol}\n delta = {dep_col_sol}\n Hd = {Hd_sol}\n")
print("Computing the solution profiles:")

I_sol, Q_sol = fs.solve_profiles(a_sol, r_sol, eps_sol, dep_col_sol, Hd_sol)
I_sol = I_sol[-1,:,mu].copy()
Q_sol = Q_sol[-1,:,mu].copy()

if(np.min(I_sol) < 0):
    print('Bad solution parameters, stopping.')
    exit()

### AD NOISE TO THE "OBSERVED" PROFILE
I_sol = add_noise(I_sol, std)
Q_sol = add_noise(Q_sol, std)

################     COMPUTE THE INITIAL GUESS PROFILES AND RADIATION FIELD      ###############
print("\nInitial parameters: ")
print(f" a = {a_initial}\n r = {r_initial}\n eps = {eps_initial}\n delta = {dep_col_initial}\n Hd = {Hd_initial}\n")
##########  MINIMIZE THE CHI2 FUNCTION WITH GIVEN RANGE CONSTRAINS #########

x_0 = np.array([a_initial,r_initial,eps_initial,dep_col_initial,Hd_initial])
x_l = np.array([1e-12,1e-12,1e-4,0,0.2])
x_u = np.array([1,1,1,10,1])

points = x_0.copy()

chi2_0, I_0, Q_0 = chi2(x_0, I_sol, Q_sol, std, w_I, w_Q, mu)

for itt in range(max_itter):
# calculation of the drerivatives of the forward model
    
    print(f'itteration {itt} with a step size of {step_size}')
    print("\nNew parameters: ")
    print(f" a = {x_0[0]}\n r = {x_0[1]}\n eps = {x_0[2]}\n delta = {x_0[3]}\n Hd = {x_0[4]}\n")

    points = np.vstack((points,x_0))
    xs = surroundings(x_0, h)
    beta = compute_gradient(I_sol, Q_sol, I_0, Q_0, x_0, xs, std, w_I, w_Q, mu)

    x_1 = x_0 - step_size*beta

    for i in range(len(x_0)):
        if x_1[i] > x_u[i]:
            x_1[i] = x_u[i]
        elif x_1[i] < x_l[i]:
            x_1[i] = x_l[i]

    chi2_1, I_1, Q_1 = chi2(x_1, I_sol, Q_sol, std, w_I, w_Q, mu)

    if chi2_1 < 1e3:
        break

    x_0 = x_1.copy()
    chi2_0 = chi2_1.copy()
    I_0 = I_1.copy()
    Q_0 = Q_1.copy()

##### PRINT AND PLOT THE SOLUTION AND COMPARE IT TO THE INITIAL AND OBSERVED PROFILES ####
print("Computing the initial and final profiles:")

I_initial, Q_initial = fs.solve_profiles(a_initial, r_initial, eps_initial, dep_col_initial, Hd_initial)
I_initial = I_initial[-1,:,mu].copy()
Q_initial = Q_initial[-1,:,mu].copy()

a_res, r_res, eps_res, dep_col_res, Hd_res = x_0[0], x_0[1], x_0[2], x_0[3], x_0[4]
I_res, Q_res = fs.solve_profiles(a_res, r_res, eps_res, dep_col_res, Hd_res)
I_res = I_res[-1,:,mu].copy()
Q_res = Q_res[-1,:,mu].copy()

print('\nFound Parameters - Solution parameters:')
print('a_result     = %1.2e \t a_solution     = %1.2e \t a_initial      = %1.2e' % (a_res, a_sol, a_initial))
print('r_result     = %1.2e \t r_solution     = %1.2e \t r_initial      = %1.2e' % (r_res, r_sol, r_initial) )
print('eps_result   = %1.2e \t eps_solution   = %1.2e \t eps_initial    = %1.2e' % (eps_res, eps_sol, eps_initial) )
print('delta_result = %1.2e \t delta_solution = %1.2e \t delta_initial  = %1.2e' % (dep_col_res, dep_col_sol, dep_col_initial) )
print('Hd_result    = %1.2e \t Hd_solution    = %1.2e \t Hd_initial     = %1.2e\n' % (Hd_res, Hd_sol, Hd_initial) )
print('w_I          = %1.2e \t w_Q            = %1.2e' % (w_I, w_Q) )
print('mu           = %1.2e ' % mu )
print('std          = %1.2e ' % std )

plt.plot(I_sol, 'ok', label=r'$I/B_{\nu}$ "observed"')
plt.plot(I_initial, 'r', label=r'$I/B_{\nu}$ initial parameters')
plt.plot(I_res, 'b', label=r'$I/B_{\nu}$ inverted parameters')
plt.legend(); plt.xlabel(r'$\nu\ (Hz)$')
plt.show()
plt.plot(Q_initial, 'r--', label='$Q/I$ initial parameters')
plt.plot(Q_sol, 'ok', label='$Q/I$ "observed"')
plt.plot(Q_res, 'b--', label='$Q/I$ inverted parameters')
plt.legend(); plt.xlabel(r'$\nu\ (Hz)$')
plt.show()

plt.plot(points[:,2],points[:,4],'o-.k', markersize=3, linewidth=1)
plt.plot(points[-1,2],points[-1,4],'ok', markersize=7)
plt.plot(eps_sol,Hd_sol,'or', markersize=10)
plt.plot(eps_initial,Hd_initial,'ob', markersize=10)
plt.xlim(x_l[2],x_u[2]);    plt.ylim(x_l[4],x_u[4])
plt.title('Movement in the parameter space of the inversion vs solution')
plt.xlabel('eps');          plt.ylabel('H_d')
plt.xscale('log')
plt.show()

plt.plot(points[:,2],points[:,3],'o-.k', markersize=3, linewidth=1)
plt.plot(points[-1,2],points[-1,3],'ok', markersize=7)
plt.plot(eps_sol,dep_col_sol,'or', markersize=10)
plt.plot(eps_initial,dep_col_initial,'ob', markersize=10)
plt.xlim(x_l[2],x_u[2]);    plt.ylim(x_l[3],x_u[3])
plt.title('Movement in the parameter space of the inversion vs solution')
plt.xlabel('eps');          plt.ylabel('dep_col')
plt.xscale('log')
plt.show()

plt.plot(points[:,3],points[:,4],'o-.k', markersize=3, linewidth=1)
plt.plot(points[-1,3],points[-1,4],'ok', markersize=7)
plt.plot(dep_col_sol,Hd_sol,'or', markersize=10)
plt.plot(dep_col_initial,Hd_initial,'ob', markersize=10)
plt.xlim(x_l[3],x_u[3]);    plt.ylim(x_l[4],x_u[4])
plt.title('Movement in the parameter space of the inversion vs solution')
plt.xlabel('dep_col');      plt.ylabel('H_d')
plt.show()

plt.plot(points[:,1],points[:,2],'o-.k', markersize=3, linewidth=1)
plt.plot(points[-1,1],points[-1,2],'ok', markersize=7)
plt.plot(r_sol,eps_sol,'or', markersize=10)
plt.plot(r_initial,eps_initial,'ob', markersize=10)
plt.xlim(x_l[1],x_u[1]);    plt.ylim(x_l[2],x_u[2])
plt.title('Movement in the parameter space of the inversion vs solution')
plt.xlabel('r');            plt.ylabel('eps')
plt.xscale('log');          plt.yscale('log')
plt.show()

File: C/inversion/src/cuda/inversion_GD.c
/**********************************************************************/
/*  Program to solve the inversion of the 2 level atom problem        */
/*  Author: Andres Vicente Arevalo      Date: 19-03-2021              */
/**********************************************************************/
#include "includes_definitions.h"

// Define the "observed" parameters
const double a_sol = 1e-3;                      /* # dumping Voigt profile a=gam/(2^1/2*sig) */
const double r_sol = 1e-3;                     /* # line strength XCI/XLI */
const double eps_sol = 1e-4;                    /* # Phot. dest. probability (LTE=1,NLTE=1e-4) */
const double dep_col_sol = 0.8;                /* # Depolirarization colisions (delta) */
const double Hd_sol = 0.5;                      /* # Hanle depolarization factor [1/5, 1] */

int main(){

    int i, j, k, itt;
    double I_full[nz][nw][qnd], Q_full[nz][nw][qnd];
    double I_0[nw], Q_0[nw], I_1[nw], Q_1[nw], I_obs_sol[nw], Q_obs_sol[nw];
    double x_0[numpar], x_1[numpar], x_sol[numpar] = {a_sol, r_sol, eps_sol, dep_col_sol, Hd_sol};
    double x_l[numpar] = {1e-12, 1e-12, 1e-4, 0, 0.2};
    double x_u[numpar] = {1, 1, 1, 10 ,1};
    double step_size[numpar] = {1e-2,1e-7,1e-6,1e-1,1e-1};
    double xs[numpar][numpar], beta[numpar];
    float progres;

    double chi2_0, chi2_1, cc=1;
    int new_point = 1; 

    fprintf(stdout,"=====================================\n");
    fprintf(stdout, "\nSOLUTION PARAMETERS:\n");
    fprintf(stdout, " a = %1.8e\n r = %1.8e\n eps = %1.8e\n delta = %1.8e\n Hd = %1.8e\n\n",\
                          x_sol[0],    x_sol[1],     x_sol[2],      x_sol[3],    x_sol[4]);
    fprintf(stdout,"=====================================\n");

    fprintf(stdout, "Computing the solution profiles... \n\n");
    solve_profiles( x_sol[0], x_sol[1], x_sol[2], x_sol[3], x_sol[4], I_full, Q_full);
    for (j = 0; j < nw; j++){
        I_obs_sol[j] = I_full[nz-1][j][mu_sel];
        Q_obs_sol[j] = Q_full[nz-1][j][mu_sel];
    }
    fprintf(stdout,"=====================================\n");

    add_noise_1D(nw, I_obs_sol, STD);
    add_noise_1D(nw, Q_obs_sol, STD);

    // initialice the parameters to start the inversion
    // for (i = 0; i < numpar; i++){
    //     x_0[i] = (double)rand()/(double)(RAND_MAX) * (x_u[i] - x_l[i]) + x_l[i];
    // }
    x_0[0] = 1e-5; x_0[1] = 1e-5; x_0[2] = 1e-3;  x_0[3] = 1e-1; x_0[4] = 0.6;

    fprintf(stdout, "\nINITIAL PARAMETERS:\n");
    fprintf(stdout, " a = %1.8e\n r = %1.8e\n eps = %1.8e\n delta = %1.8e\n Hd = %1.8e\n\n",\
                          x_0[0],     x_0[1],        x_0[2],        x_0[3],       x_0[4]);
    fprintf(stdout,"=====================================\n");
    
    chi2_0 = chi2_calc(x_0, I_obs_sol, Q_obs_sol, I_0, Q_0, STD, WI, WQ);

    for (itt = 0; itt < max_iter_inversion; itt++){

        fprintf(stdout,"Step %i of %i\n",itt, max_iter_inversion);
        progres =  100.0 * itt/ max_iter_inversion ;
        fprintf(stdout, "[");
        for (i=0; i<70; i++) {
            if ( i < progres*7/10 ) {
                fprintf(stdout, "#");
            }else{
                fprintf(stdout, "-");
            }
        }
        fprintf(stdout, "] %1.2f %%\n", progres);

        if(new_point){
            surroundings_calc(x_0, xs, HH);
            check_range(x_0, xs, x_l, x_u);
            compute_gradient(x_0, xs, beta, I_obs_sol, Q_obs_sol, HH, STD, WI, WQ);
        }

        for (i = 0; i < numpar; i++){
            x_1[i] = x_0[i] - step_size[i]*beta[i]*cc;
        }


        for (i=0; i<numpar; i++){
            if (x_1[i] > x_u[i]){
                x_1[i] = x_u[i];
            }else if (x_1[i] < x_l[i]){
                x_1[i] = x_l[i];
            }
        }
       
        chi2_1 = chi2_calc(x_1, I_obs_sol, Q_obs_sol, I_1, Q_1, STD, WI, WQ);
        fprintf(stdout,"Total Chi^2 %1.6e\n with step parameter of %1.3e\n",chi2_1,cc);
        
        // If we have very good loss stop
        if (chi2_1 < 10){
            break;
        }

        if (chi2_1 <= chi2_0 || itt == 0){
            for (i = 0; i < numpar; i++){
                x_0[i] = x_1[i];
            }
            cc = cc*1.25;
            new_point = 1;
            chi2_0 = chi2_1;

            for (j = 0; j < nw; j++){
                I_0[j] = I_1[j];
                Q_0[j] = Q_1[j];
            }

            fprintf(stdout,"=====================================\n");
            fprintf(stdout, "\nNew parameters:\n");
            fprintf(stdout, " a = %1.8e\n r = %1.8e\n eps = %1.8e\n delta = %1.8e\n Hd = %1.8e\n\n",\
                                x_0[0], x_0[1], x_0[2], x_0[3], x_0[4]);
            fprintf(stdout,"=====================================\n");
        }else{
            cc = cc/2;            
            new_point = 0;
        }
    }
    
    fprintf(stdout, "\nFINISHED AFTER %i ITTERATIONS\n", itt);
    fprintf(stdout, "\n--------------------------------------------------------------------\n");
    fprintf(stdout, "\tINVERTED PARAMETERS\t | \tSOLUTION PARAMETERS\t\n");
    fprintf(stdout, "--------------------------------------------------------------------\n");
    fprintf(stdout, "  a = %1.8e \t\t |  a_sol = %1.8e\n  r = %1.8e \t\t |  r_sol = %1.8e\
    \n  eps = %1.8e \t\t |  eps_sol = %1.8e\n  delta = %1.8e \t |  delta_sol = %1.8e\
    \n  Hd = %1.8e \t\t |  Hd_sol = %1.8e\n",\
    x_0[0], x_sol[0], x_0[1], x_sol[1], x_0[2], x_sol[2], x_0[3], x_sol[3], x_0[4], x_sol[4]);
    fprintf(stdout, "--------------------------------------------------------------------\n");
    
    /* define array, counter, and file name, and open the file */

    FILE *fp;
    fp=fopen("./figures/profiles.txt","w");
    
    fprintf(fp, "# index, I_solution, Q_solution, I_inverted, Q_inverted\n");
    for(i = 0; i < nw; i++){
       fprintf (fp, " %i , %1.8e , %1.8e , %1.8e , %1.8e\n",\
                    i, I_obs_sol[i], Q_obs_sol[i], I_0[i], Q_0[i]);
    }
    fclose (fp);

    fp=fopen("./figures/parameters_inverted.txt","w");

    fprintf(fp, "\nFINISHED AFTER %i ITTERATIONS\n", itt);
    fprintf(fp, "\n------------------------------------------------------------------------\n");
    fprintf(fp, "\tINVERTED PARAMETERS\t | \tSOLUTION PARAMETERS\t\n");
    fprintf(fp, "------------------------------------------------------------------------\n");
    fprintf(fp, "  a = %1.8e \t\t\t |  a_sol = %1.8e\n  r = %1.8e \t\t\t |  r_sol = %1.8e\
    \n  eps = %1.8e \t\t\t |  eps_sol = %1.8e\n  delta = %1.8e \t\t |  delta_sol = %1.8e\
    \n  Hd = %1.8e \t\t\t |  Hd_sol = %1.8e\n",\
    x_0[0], x_sol[0], x_0[1], x_sol[1], x_0[2], x_sol[2], x_0[3], x_sol[3], x_0[4], x_sol[4]);
    fprintf(fp, "------------------------------------------------------------------------\n");
    fprintf(fp, "\t\tEXECUTION PARAMETERS\n");
    fprintf(fp, "------------------------------------------------------------------------\n");
    fprintf(fp, " Maximum iterrations = %i\n Noise of the profiles = %f\
                \n Selected direction of observation = %i\n step of the derivative = %f\
                \n weigths:\t w_I = %f\t w_Q = %f\n",max_iter_inversion,STD,mu_sel,HH,WI,WQ);
    fprintf(fp, "------------------------------------------------------------------------\n");
    fclose (fp);

    return 0;
}
